Issue #11446: Replace verify method in ImportControlCheckTest

arinmodi opened the pull request:
Subtask of #11446  : Update The ImportControlCheckTest

arinmodi commented:
@romani,

![image](https://user-images.githubusercontent.com/61725413/201063735-65549aab-a2b9-4253-9a86-d14e9692cc4b.png)


what's this error ?

arinmodi commented:
> Close. It is actually saying "Violation was expected at 11 but no violation was found instead". The `but was: []` shows an empty array which lists all the actual violations. The top text right above the `---` as says "1 item was missing and it is line 11".

how should I solve this ? I tried following things,

- assign the expected array to CommanUtils.Empty_Array, but now worked
- use execute method with configs but that generated the pml violation error,

can you please throw some hints

rnveach commented:
The test in question is `testFileNameNoExtension`. You are getting an error on expected violations, so that means the checkstyle run is happening but it is not giving the results you expected. To me, this makes me believe it is either a configuration issue (the checks and properties being used for the test) or an input file issue (the file changed in such a way to no longer produce violations). I would start looking there.

rnveach commented:
Debug and figure that out. Ask what did you change to cause this behavior.

arinmodi commented:
> Debug and figure that out. Ask what did you change to cause this behavior.

I found that it is something related to the treeWalker config, can I set that in input file ?

arinmodi commented:
@rnveach, can I specify the treeWalker property in testFileNameNoExtension's comment ?

arinmodi commented:
> @rnveach, can I specify the treeWalker property in testFileNameNoExtension's comment ?

@rnveach  ?

nrmancuso commented:
@arinmodi please stop spamming maintainers, we will come to all comments/PRs eventually

arinmodi commented:
okay, sorry for that

rnveach commented:
You should be able to. Look at other input files using our BDD.

nrmancuso commented:
@arinmodi try to grep in the test input directory by the property you want to set

arinmodi commented:
can any maintainer explain me following test,

1. FileLengthCheckTest : In this test, files inline configs are parsed and added as child of checker,
2. ImportControlCheckTest : In this test files, inline configs are parsed and added as child of TreeWalker.

Both the test file using same method for verification, how this is achieved ?

arinmodi commented:
I think currently there is no method that allow us to add property of TreeWalker

rnveach commented:
> currently there is no method that allow us to add property of TreeWalker

Show us what is the issue/failure with adding TreeWalker to the BDD similar to the others.
Example: https://github.com/checkstyle/checkstyle/commit/e662bd96c3527104fc57a91a130bf74e031c106c#diff-b6e359d604e2d75f3b32693812402279e04383f5e3ab5061a810689d864b822eR1

arinmodi commented:
> > currently there is no method that allow us to add property of TreeWalker
> 
> Show us what is the issue/failure with adding TreeWalker to the BDD similar to the others. Example: [e662bd9#diff-b6e359d604e2d75f3b32693812402279e04383f5e3ab5061a810689d864b822eR1](https://github.com/checkstyle/checkstyle/commit/e662bd96c3527104fc57a91a130bf74e031c106c#diff-b6e359d604e2d75f3b32693812402279e04383f5e3ab5061a810689d864b822eR1)

> > currently there is no method that allow us to add property of TreeWalker
> 
> Show us what is the issue/failure with adding TreeWalker to the BDD similar to the others. Example: [e662bd9#diff-b6e359d604e2d75f3b32693812402279e04383f5e3ab5061a810689d864b822eR1](https://github.com/checkstyle/checkstyle/commit/e662bd96c3527104fc57a91a130bf74e031c106c#diff-b6e359d604e2d75f3b32693812402279e04383f5e3ab5061a810689d864b822eR1)

TreeWalker can't be added as child of the TreeWalker

arinmodi commented:
![image](https://user-images.githubusercontent.com/61725413/201390843-a0e48c11-c154-4b93-869c-6b340465a2f2.png)

![image](https://user-images.githubusercontent.com/61725413/201391574-02d3d83d-32e3-4865-980b-7b9aacaaefd3.png)

Error Message:

```
TreeWalker is not allowed as a parent of com.puppycrawl.tools.checkstyle.TreeWalker Please review 'Parent Module' section for this Check in web documentation if Check is standard.
```

File Content : 

```
/*
com.puppycrawl.tools.checkstyle.TreeWalker
fileExtensions = null


ImportControl
file = (file)InputImportControlFileNameNoExtension.xml
path = (default).*


*/

package com.puppycrawl.tools.checkstyle.checks.imports.importcontrol;

import java.awt.Image; // violation 'Disallowed import - java.awt.Image.'

public @interface InputImportControlFileNameNoExtension {
}
```

romani commented:
We do not need `com.puppycrawl.tools.checkstyle.TreeWalker` in input file, test code knows all parents and should place it properly.

romani commented on src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java:
```diff
@@ -231,8 +233,32 @@ protected final void verifyWithInlineConfigParser(String filePath, String... exp
+     * @throws Exception if exception occurs during verification process.
+     */
+    protected final void verifyWithInlineConfigParser(String filePath,
+                                                      DefaultConfiguration treeWalker,
```
Non of other execution require this, but almost all of them are children of Treewalker
Please debug around https://github.com/checkstyle/checkstyle/blob/c47dc6e5b4f95db935a391ea527c4a0300b10ad3/src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java#L119
To figure out why it works for others but for your Check

arinmodi replied:
I debug and found that currently there is no check that setting the property of the treeWalker.

romani replied:
Ok I see it 
https://github.com/checkstyle/checkstyle/pull/12386/files#diff-74a37ae26b9b5489dc1aeb5346ee75381aa4291455c9b6e488f56ccd1c5d90d2L430

romani replied:
But in this case we need something like https://github.com/checkstyle/checkstyle/pull/12386#issuecomment-1311950902

As all should be defined in Input file.
Please create one more PR and try to implement not adding traleewaler under Treewalker but to update properties of it from values from file.

arinmodi replied:
ok

arinmodi replied:
can you please create the Issue for this ?

arinmodi replied:
I am having some problem in handling the exception, what should I write in catch,

here is my function for processing each moduleconfig,

```
    public DefaultConfiguration createConfiguration() {
        final DefaultConfiguration root = new DefaultConfiguration(ROOT_MODULE_NAME);
        final DefaultConfiguration treeWalker =
                new DefaultConfiguration(TreeWalker.class.getName());
        root.addProperty("charset", StandardCharsets.UTF_8.name());
        childrenModules
                .stream()
                .map(ModuleInputConfiguration::createConfiguration)
                .forEach(moduleConfig -> {
                    processModuleConfigs(root, treeWalker, moduleConfig);
                });
        root.addChild(treeWalker);
        return root;
    }

    private static void processModuleConfigs(DefaultConfiguration root,
                                     DefaultConfiguration treeWalker,
                                     DefaultConfiguration moduleConfig) {
        if (CHECKER_CHILDREN.contains(moduleConfig.getName())) {
            root.addChild(moduleConfig);
        }
        else if (TREE_WALKER_MODULE_NAME.equals(moduleConfig.getName())) {
            for (String property : moduleConfig.getPropertyNames()) {
                try {
                    treeWalker.addProperty(property, moduleConfig.getProperty(property));
                } catch (CheckstyleException ex) {
                    
                }
            }
        }
        else {
            treeWalker.addChild(moduleConfig);
        }
    }
```

romani replied:
Most likely this PR will win https://github.com/checkstyle/checkstyle/pull/12395

romani reviewed (changes requested):
Items

arinmodi commented:
now, how do I processed with this PR ?

rnveach commented:
@arinmodi If you only have 1 test to modify and it requires code like https://github.com/checkstyle/checkstyle/pull/12395 then this PR is blocked until the other PR is merged. Your free to take up another test or issue in the meantime, but I don't see the other PR being too much longer.

rnveach commented:
@arinmodi Other PR was merged. Please rebase and finish. Let us know if there is any other issues.

arinmodi commented:
can anyone explain me why this is failing "Check no closed issue references" ?

nrmancuso commented:
> can anyone explain me why this is failing "Check no closed issue references" ?

fix is at https://github.com/checkstyle/checkstyle/pull/12414

rnveach commented:
@arinmodi I merged a fix.

arinmodi commented:
> @arinmodi I merged a fix.

Still failing, I did rebase

romani reviewed (approved):
ok to merge, we will restart CI to make it green.
